xl/xenops_server_xenlight.ml

```diff
@@ -2995,9 +3021,11 @@ let process_one_watch xc xs (path, token) =
 	if path = _introduceDomain || path = _releaseDomain
 	then look_for_different_domains xc xs
 	else match List.filter (fun x -> x <> "") (Xstringext.String.split '/' path) with
-		| "local" :: "domain" :: domid :: "backend" :: kind :: frontend :: devid :: _ ->
+		| "local" :: "domain" :: domid :: "backend" :: kind :: frontend :: devid :: key ->
 			debug "Watch on backend domid: %s kind: %s -> frontend domid: %s devid: %s" domid kind frontend devid;
-			fire_event_on_device frontend kind devid
+			fire_event_on_device frontend kind devid;
+			(* If this event was a state change then this might be the first time we see evidence of PV drivers *)
+			if key = ["state"] then maybe_update_pv_drivers_detected ~xc ~xs (int_of_string frontend) path
```

Comment:
The other check I mentioned is here https://github.com/xapi-project/ocaml-xen-lowlevel-libs/blob/master/lib/xenctrl_stubs.c#L1250.

It checks that the guest in the domain has registerd the CALLBACK_IRQ.

It's wrapped up in the Xenctrl library here https://github.com/xapi-project/ocaml-xen-lowlevel-libs/blob/master/lib/xenctrl.mli#L309

and it's used here https://github.com/xapi-project/xenopsd/blob/master/xc/domain.ml#L336 by Xenopsd when decided whether it should wait for cooperation from the guest when performing a shutdown.

I realise from our converstaion that you're wanting to distinguish between that and specifically PV device drivers so that's fine. Just thought I'd follow up on what we talked about.

I'm happy with the semantics of the field as they are in this pull request though.